spec.md

```diff
@@ -2734,6 +2780,254 @@ message NodeExpandVolumeResponse {
 | Volume in use | 9 FAILED_PRECONDITION | Indicates that the volume corresponding to the specified `volume_id` could not be expanded because it is node-published or node-staged and the underlying filesystem does not support expansion of published or staged volumes. | Caller MUST NOT retry. |
 | Unsupported capacity_range | 11 OUT_OF_RANGE | Indicates that the capacity range is not allowed by the Plugin. More human-readable information MAY be provided in the gRPC `status.message` field. | Caller MUST fix the capacity range before retrying. |
 
+### Group Controller Service RPCs
+
+#### `GroupControllerGetCapabilities`
+
+A GroupController Plugin MUST implement this RPC call. This RPC allows the CO to check the supported capabilities of group controller service provided by the Plugin.
+
+```protobuf
+message GroupControllerGetCapabilitiesRequest {
+  // Intentionally empty.
+}
+
+message GroupControllerGetCapabilitiesResponse {
+  // All the capabilities that the group controller service supports.
+  // This field is OPTIONAL.
+  repeated GroupControllerServiceCapability capabilities = 1;
+}
+
+// Specifies a capability of the group controller service.
+message GroupControllerServiceCapability {
+  message RPC {
+    enum Type {
+      UNKNOWN = 0;
+
+      // Indicates that the group controller plugin supports
+      // creating, deleting, and getting details of a volume
+      // group snapshot.
+      CREATE_DELETE_GET_VOLUME_GROUP_SNAPSHOT = 1
+      [(alpha_enum_value) = true];
+    }
+
+    Type type = 1;
+  }
+
+  oneof type {
+    // RPC that the controller supports.
+    RPC rpc = 1;
+  }
+}
+```
+
+##### GroupControllerGetCapabilities Errors
+
+If the plugin is unable to complete the GroupControllerGetCapabilities call successfully, it MUST return a non-ok gRPC code in the gRPC status.
+
+#### `CreateVolumeGroupSnapshot`
+
+**ALPHA FEATURE**
+
+A Group Controller Plugin MUST implement this RPC call if it has `CREATE_DELETE_GET_VOLUME_GROUP_SNAPSHOT` controller capability.
+This RPC will be called by the CO to create a new volume group snapshot from a list of source volumes on behalf of a user.
```

Comment:
Should we add a consistency requirement (or recommendation) for created snapshots?

Reply:
Added this: This group snapshot SHOULD be crash consistent.

Reply:
That wording is still a big vague. Sorry I overlooked this part of the spec during my review, but it IS a very important aspect of this spec to explain why a group snapshot is not functionally identical to a bunch of snapshots taken close to the same time.

I propose this wording:

The purpose of this call is to request the creation of a multi-volume (group) snapshot. This group snapshot MUST give a write-order consistency guarantee or fail if that's not possible. That is to say, all the of the volume snapshots in the group must be taken at the same point in time relative to a stream of write traffic to the specified volumes. Note that calls to this function MUST be idempotent - the function may be called multiple times for the same name - the group snapshot must only be created once.


I think that we have to use MUST rather than SHOULD with the consistency guarantee to rule out implementations that can't offer the needed guarantee -- otherwise we poison this feature for end users who actually want to depend on the guarantee. SHOULD leaves room that the guarantee is not met and will harm adoption.